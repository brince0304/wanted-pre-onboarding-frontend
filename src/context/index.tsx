import React, {
    createContext, Dispatch,
    useContext, useEffect,
    useReducer,

} from "react";

type Action = { type: "SET_TOKEN"; token: string | null; };

type State = { accessToken: string | null; }

type AppDispatch = Dispatch<Action>;

const initialState: State = {
    accessToken: null,
}

const stateName = 'tokenState';
const TokenStateContext = createContext<State | null>(null);
const TokenDispatchContext = createContext<AppDispatch | null>(null);

const TokenReducer = (state: State, action: Action) => {
    switch (action.type) {
        case "SET_TOKEN":
            return {
                ...state,
                accessToken: action.token
            };
        default:
            throw new Error("Unhandled action");
    }
};

export function TokenProvider({children}: { children: React.ReactNode }) {

    const [state, dispatch] = useReducer(TokenReducer, initialState);

    useEffect(() => {
        // 로컬 스토리지에서 상태 값 불러오기
        const storedState = localStorage.getItem(stateName);
        if (storedState) {
            setToken(dispatch, JSON.parse(storedState).accessToken);
        }
    }, []);

    useEffect(() => {
        // 상태 값 로컬 스토리지에 저장하기
        localStorage.setItem(stateName, JSON.stringify(state));
    }, [state]);


    return (
        <TokenStateContext.Provider value={state}>
            <TokenDispatchContext.Provider value={dispatch}>
                {children}
            </TokenDispatchContext.Provider>
        </TokenStateContext.Provider>
    );
}

export function useTokenState() {
    const state = useContext(TokenStateContext);
    if (!state) {
        throw new Error("Connot find TokenStateProvider");
    }
    return state;
}

export function useTokenDispatch() {
    const dispatch = useContext(TokenDispatchContext);
    if (!dispatch) {
        throw new Error("Cannot find DispatchProvider");
    }
    return dispatch;
}

export const setToken = (dispatch: AppDispatch, token: string) => {
    dispatch({type: "SET_TOKEN", token});
}

export const getTokenFromLocalStorage = () => {
    const tokenState = localStorage.getItem(stateName);
    if (tokenState) {
        return JSON.parse(tokenState).accessToken;
    }
    return null;
}

export const removeTokenFromLocalStorage = () => {
    localStorage.removeItem(stateName);
}
