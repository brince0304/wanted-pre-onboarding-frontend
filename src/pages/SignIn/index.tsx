import Box from "@mui/material/Box";
import React, {FormEvent, useCallback, useContext, useEffect} from "react";
import {Button, Typography} from "@mui/material";
import {useNavigate} from "react-router";
import {useFormControl} from "../../hooks/useFormControl";
import ValidationInput from "../../components/ValidationInput";
import styled from "@emotion/styled";
import {postSignin} from "../../apis";
import { setToken, useTokenDispatch} from "../../context";

const StyledFormControl = styled.form`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 100%;
  gap: 20px;
`;

const StyledInputBox = styled(Box)`
  width: 100%;
  height: 100%;
`;

const StyledSignInButton = styled(Button)`
  width: 100%;
  height: 100%;
`;

const SignIn = () => {
    const dispatch = useTokenDispatch();
    const navigate = useNavigate();
    const emailRegex = /@/;
    const passwordRegex = /.{8,}/;
    const [onChangeEmail, email,setEmail, emailValidation, setEmailValidation] = useFormControl({regex: emailRegex});
    const [onChangePassword, password,setPassword, passwordValidation, setPasswordValidation] = useFormControl({regex: passwordRegex});
    const [emailValidationMessage, setEmailValidationMessage] = React.useState<string>("이메일 형식이 아닙니다.");
    const [passwordValidationMessage, setPasswordValidationMessage] = React.useState<string>("비밀번호는 8자리 이상이어야 합니다.");

    const emailInputProps = {
        testId: "email-input",
        validationMessage: emailValidationMessage,
        validationTrueMessage: "이메일 형식이 올바릅니다.",
        placeholder: "이메일을 입력해주세요.",
        type: "email",
        value: email,
        onChange: onChangeEmail,
        validation: emailValidation,
        label: "이메일"
    };

    const passwordInputProps = {
        testId: "password-input",
        validationMessage: passwordValidationMessage,
        validationTrueMessage: "정상적인 비밀번호 입니다.",
        placeholder: "비밀번호를 입력해주세요.",
        type: "password",
        value: password,
        onChange: onChangePassword,
        validation: passwordValidation,
        label: "비밀번호"
    };

    const handleEmailValidationFalse = () => {
        setEmailValidation(false);
        setEmailValidationMessage("해당 사용자가 존재하지 않습니다.");
    }

    const handlePasswordValidationFalse = () => {
        setPasswordValidation(false);
        setPasswordValidationMessage("비밀번호가 일치하지 않습니다.");
    }

    const onSubmit = (e: FormEvent<HTMLFormElement>) => {
        e.preventDefault();
        if (emailValidation && passwordValidation) {
            postSignin({email, password}).then((res) => {
                if (res) {
                    setToken(dispatch, res.access_token);
                    navigate("/todo");
                }
            }).catch((err) => {
                    if (err.response.status === 404) {
                        handleEmailValidationFalse();
                    } else if (err.response.status === 401) {
                        handlePasswordValidationFalse();
                    }
                }
            );
        }
    };

    return (
        <StyledFormControl onSubmit={onSubmit}>
            로그인
            <StyledInputBox>
                <ValidationInput {...emailInputProps}/>
            </StyledInputBox>
            <StyledInputBox>
                <ValidationInput {...passwordInputProps}/>
            </StyledInputBox>
            <StyledSignInButton type={"submit"} disabled={!emailValidation || !passwordValidation}
                                variant={"contained"}
                                data-testid={"signin-button"}>
                로그인
            </StyledSignInButton>
        </StyledFormControl>
    );
};

export default SignIn;
