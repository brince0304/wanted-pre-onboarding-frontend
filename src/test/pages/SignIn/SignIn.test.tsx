import SignIn from "../../../pages/SignIn";
import {fireEvent, render, screen, waitFor} from "@testing-library/react";
import  {TokenProvider} from "../../../context";
import React, {ReactNode} from "react";
import {BrowserRouter} from "react-router-dom";



const apiMethods = require("../../../apis/index.ts");
const context = require("../../../context/index.tsx");

const mockedDispatch = jest.fn();
jest.mock('../../../context', () => ({
    ...jest.requireActual('../../../context') as any,
    useTokenDispatch: () => mockedDispatch,
}));

const mockedUsedNavigate = jest.fn();

jest.mock("react-router", () => ({
    ...jest.requireActual("react-router"),
    useNavigate: () => mockedUsedNavigate
}));


beforeEach(() => {
    mockedUsedNavigate.mockReset();
    mockedDispatch.mockReset();
});

afterEach(() => {
    jest.clearAllMocks();
    jest.restoreAllMocks();
});


function customRender(children: ReactNode) {
    return render(
        <TokenProvider>
            <BrowserRouter>
                {children}
            </BrowserRouter>
        </TokenProvider>
    );
}

beforeEach(() => {
    mockedDispatch.mockReset();
    mockedUsedNavigate.mockReset();
});

describe("<SignIn/>", () => {
    it("matches snapshot", () => {
        const utils = customRender(<SignIn/>);
        expect(utils.container).toMatchSnapshot();
    });
    it("이메일, 비밀번호가 입력이 안되어있을 때 버튼 disabled", () => {
        customRender(<SignIn/>);
        const signInButton = screen.getByTestId("signin-button");
        const emailInput = screen.getByTestId("email-input");
        const passwordInput = screen.getByTestId("password-input");
        fireEvent.change(emailInput, {target: {value: ""}});
        fireEvent.change(passwordInput, {target: {value: ""}});
        expect(signInButton).toBeDisabled();
    });
    it("이메일, 비밀번호가 조건에 맞게 입력이 되어있을 때 버튼 disabled 해제", () => {
        customRender(<SignIn/>);
        const signInButton = screen.getByTestId("signin-button");
        const emailInput = screen.getByTestId("email-input");
        const passwordInput = screen.getByTestId("password-input");
        fireEvent.change(emailInput, {target: {value: "emai9090l@email.email"}});
        fireEvent.change(passwordInput, {target: {value: "password"}});
        expect(signInButton).not.toBeDisabled();
    });
    it("이메일이 옳지 않을 때 validation 에 메시지 추가", async () => {
        const postSignIn = jest.spyOn(apiMethods, "postSignin").mockRejectedValue({
            response: {
                status: 404,
                data: {message: "해당 사용자가 존재하지 않습니다."}
            }
        });
        customRender(<SignIn/>);
        const signInButton = screen.getByTestId("signin-button");
        const emailInput = screen.getByTestId("email-input");
        const passwordInput = screen.getByTestId("password-input");
        fireEvent.change(emailInput, {target: {value: "email@email.email"}});
        fireEvent.change(passwordInput, {target: {value: "password"}});
        fireEvent.click(signInButton);
        await waitFor(() => {
            const validationIcon = screen.getAllByTestId('CheckCircleIcon')[0];
            expect(validationIcon.getAttribute('aria-label')).toBe('해당 사용자가 존재하지 않습니다.');
        });
    });
    it("비밀번호가 옳지 않을 때 validation 에 메시지 추가", async () => {
        const postSignIn = jest.spyOn(apiMethods, "postSignin").mockRejectedValue({
            response: {
                status: 401,
                data: {message: "Unauthorized"}
            }
        });
        customRender(<SignIn/>);
        const signInButton = screen.getByTestId("signin-button");
        const emailInput = screen.getByTestId("email-input");
        const passwordInput = screen.getByTestId("password-input");
        fireEvent.change(emailInput, {target: {value: "mytestemail@email.email"}});
        fireEvent.change(passwordInput, {target: {value: "password0"}});
        fireEvent.click(signInButton);
        await waitFor(() => {
            const validationIcon = screen.getAllByTestId('CheckCircleIcon')[1];
            expect(validationIcon.getAttribute('aria-label')).toBe('비밀번호가 일치하지 않습니다.');
        });
    });
    it("이메일과 비밀번호가 일치할 때 contex에 토큰 저장 후 TODO 로 이동", async () => {
        const postSignIn = jest.spyOn(apiMethods, "postSignin").mockResolvedValue({
            access_token : 'token'
        });
        const mockedSetToken = jest.spyOn(context, "setToken");
        const token = 'token';
        customRender(<SignIn/>);
        const signInButton = screen.getByTestId("signin-button");
        const emailInput = screen.getByTestId("email-input") as HTMLInputElement;
        const passwordInput = screen.getByTestId("password-input") as HTMLInputElement;
        fireEvent.change(emailInput, {target: {value: "mytestemail@email.email"}});
        fireEvent.change(passwordInput, {target: {value: "password"}});
        fireEvent.click(signInButton);
        await waitFor(() => {
            expect(postSignIn).toBeCalledTimes(1);
        });
        await waitFor(()=>{
            expect(postSignIn).toBeCalledWith({
                email: emailInput.value,
                password: passwordInput.value
            });
        })
        expect(mockedSetToken).toBeCalledTimes(1);
        expect(mockedSetToken).toBeCalledWith(mockedDispatch, token);
        expect(mockedUsedNavigate).toBeCalledWith('/todo');
    });
});
