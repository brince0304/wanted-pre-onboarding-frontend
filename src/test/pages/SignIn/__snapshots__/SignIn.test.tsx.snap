// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignIn/> matches snapshot 1`] = `
<div>
  <form
    class="css-aemo20"
  >
    로그인
    <div
      class="MuiBox-root css-irtdga"
    >
      <div
        class="MuiBox-root css-1jke3r3"
        id="validation-form"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-uvscsy-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="validation-input"
            id="validation-input-label"
          >
            이메일
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-v4u5dn-MuiInputBase-root-MuiInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
              data-testid="email-input"
              id="validation-input"
              placeholder="이메일을 입력해주세요."
              type="email"
              value=""
            />
          </div>
        </div>
        <svg
          aria-hidden="true"
          aria-label=""
          class="MuiSvgIcon-root MuiSvgIcon-colorDisabled MuiSvgIcon-fontSizeMedium  css-y4srf3-MuiSvgIcon-root"
          data-mui-internal-clone-element="true"
          data-testid="CheckCircleIcon"
          focusable="false"
          id="id-email-input-validate"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
          />
        </svg>
      </div>
    </div>
    <div
      class="MuiBox-root css-irtdga"
    >
      <div
        class="MuiBox-root css-1jke3r3"
        id="validation-form"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-uvscsy-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-standard css-aqpgxn-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="validation-input"
            id="validation-input-label"
          >
            비밀번호
          </label>
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-v4u5dn-MuiInputBase-root-MuiInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
              data-testid="password-input"
              id="validation-input"
              placeholder="비밀번호를 입력해주세요."
              type="password"
              value=""
            />
          </div>
        </div>
        <svg
          aria-hidden="true"
          aria-label=""
          class="MuiSvgIcon-root MuiSvgIcon-colorDisabled MuiSvgIcon-fontSizeMedium  css-y4srf3-MuiSvgIcon-root"
          data-mui-internal-clone-element="true"
          data-testid="CheckCircleIcon"
          focusable="false"
          id="id-password-input-validate"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
          />
        </svg>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1gbpmz4-MuiButtonBase-root-MuiButton-root"
      data-testid="signin-button"
      disabled=""
      tabindex="-1"
      type="submit"
    >
      로그인
    </button>
  </form>
</div>
`;
